nupic/research/frameworks/pytorch/imagenet/mixins/quantization_aware.py

```diff
@@ -0,0 +1,201 @@
+        )
+        self.when_disable_observers = config.get("when_disable_observers", .95)
+        self.when_freeze_batch_norm = config.get("when_freeze_batch_norm", .98)
+        self.fuse_relu = config.get("fuse_relu", False)
+
+        super().setup_experiment(config)
+
+    def transform_model(self):
+        """Prepare model for quantization"""
+
+        # prepare model for qat
+        prepare_for_qat(self.model, self.quantize_weights_per_channel, self.fuse_relu)
+        self.model.to(self.device)
+
+        # enable observers and fake quantizers to prepare for training
+        self.model.apply(enable_observer)
+        self.model.apply(enable_fake_quant)
+        self.observer_disabled = False
+        self.batch_norm_frozen = False
+
+    def pre_batch(self, model, batch_idx):
+
+        # freeze observer parameters for the last 500 batches of last epoch
+        if (not self.observer_disabled
+           and self.current_epoch == self.epochs
+           and self.batch_idx > (self.when_disable_observers * self.total_batches)):
+            self.logger.info(f"Freezing observer at epoch {self.current_epoch} "
+                             "and batch {self.batch_idx}")
+            self.model.apply(disable_observer)
+            self.observer_disabled = True
+
+        # freeze BN parameters for the last 200 batches of last epoch
+        if (not self.batch_norm_frozen
+           and self.current_epoch == self.epochs
+           and self.batch_idx > (self.when_freeze_batch_norm * self.total_batches)):
+            self.logger.info(f"Freezing BN parameters at epoch {self.current_epoch}"
+                             "and batch {self.batch_idx}")
+            self.model.apply(nni.qat.freeze_bn_stats)
+            self.batch_norm_frozen = True
+
+    @classmethod
+    def get_execution_order(cls):
+        eo = super().get_execution_order()
+        eo["setup_experiment"].append("Initialize extra variables for QAT")
+        eo["transform_model"].append("Prepare model for Quantization")
+        eo["pre_batch"].append("Freezes observers and batch norm parameters in QAT")
+        return eo
+
+
+def prepare_for_qat(model, quantize_weights_per_channel, fuse_relu):
+    """Prepares model for quantization aware training"""
+
+    # fuse models
+    model.fuse_model(fuse_relu=fuse_relu)
+
+    # set qconfig
+    if quantize_weights_per_channel:
+        qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
+    else:
+        print("Quantizating weights per tensor")
+        qconfig = QConfig(
+            activation=FakeQuantize.with_args(observer=MovingAverageMinMaxObserver,
+                                              quant_min=0,
+                                              quant_max=255,
+                                              reduce_range=True),
+            weight=default_weight_fake_quant
+        )
+    model.qconfig = qconfig
+
+    # equivalent to quantize.prepare, inplace. require for custom white list
```

Comment:
This was fixed in pytorch 1.6. You can now pass custom white list to `torch.prepare`

Reply:
Not `prepare_qat`, they only updated `prepare` in 1.6